File: examples/android/CMeSdkJava/app/src/main/java/me/connect/sdk/java/sample/messages/ProofDataHolder.java
package me.connect.sdk.java.sample.messages;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;

import me.connect.sdk.java.message.Message;

public class ProofDataHolder {
    public String threadId;
    public String name;
    public String attributes;
    public String proofReq;

    public ProofDataHolder(String threadId, String name, String attributes, String proofReq) {
        this.threadId = threadId;
        this.name = name;
        this.attributes = attributes;
        this.proofReq = proofReq;
    }

    public static ProofDataHolder extractRequestedFieldsFromProofMessage(Message msg) {
        try {
            JSONObject json = new JSONObject(msg.getPayload());
            JSONObject data = json.getJSONObject("proof_request_data");
            String threadId = json.getString("thread_id");
            String name = data.getString("name");
            JSONObject requestedAttrs = data.getJSONObject("requested_attributes");
            Iterator<String> keys = requestedAttrs.keys();
            StringBuilder attributes = new StringBuilder();
            while (keys.hasNext()) {
                String key = keys.next();
                String value = requestedAttrs.getJSONObject(key).getString("name");
                attributes.append(value);
                if (keys.hasNext()) {
                    attributes.append(", ");
                }
            }
            return new ProofDataHolder(threadId, name, attributes.toString(), msg.getPayload());
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: examples/android/CMeSdkJava/app_kt/src/main/kotlin/me/connect/sdk/java/samplekt/messages/ProofDataHolder.kt
package me.connect.sdk.java.samplekt.messages

import me.connect.sdk.java.message.Message
import org.json.JSONException
import org.json.JSONObject

class ProofDataHolder(var threadId: String, var name: String, var attributes: String, var proofReq: String) {
    companion object {
        fun extractRequestedFieldsFromProofMessage(msg: Message): ProofDataHolder? {
            return try {
                val json = JSONObject(msg.payload)
                val data = json.getJSONObject("proof_request_data")
                val threadId = json.getString("thread_id")
                val name = data.getString("name")
                val requestedAttrs = data.getJSONObject("requested_attributes")
                val keys = requestedAttrs.keys()
                val attributes = StringBuilder()
                while (keys.hasNext()) {
                    val key = keys.next()
                    val value = requestedAttrs.getJSONObject(key).getString("name")
                    attributes.append(value)
                    if (keys.hasNext()) {
                        attributes.append(", ")
                    }
                }
                ProofDataHolder(threadId, name, attributes.toString(), msg.payload)
            } catch (e: JSONException) {
                e.printStackTrace()
                null
            }
        }
    }
}
